api/api.go

```diff
@@ -1075,3 +1088,44 @@ func (api *Server) getLogsInBlock(filter *LogFilter, start, count uint64) ([]*io
 	}
 	return logs, nil
 }
+
+func (api *Server) estimateActionGasConsumptionForExecution(exec *iotextypes.Execution, sender string) (*iotexapi.EstimateActionGasConsumptionResponse, error) {
+	sc := &action.Execution{}
+	if err := sc.LoadProto(exec); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	callerAddr, err := address.FromString(sender)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	caller, err := api.bc.StateByAddr(sender)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	sc, _ = action.NewExecution(
+		sc.Contract(),
+		caller.Nonce+1,
+		sc.Amount(),
+		api.cfg.Genesis.BlockGasLimit,
+		big.NewInt(0),
+		sc.Data(),
+	)
+
+	_, receipt, err := api.bc.ExecuteContractRead(callerAddr, sc)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if receipt.Status != action.SuccessReceiptStatus {
+		return nil, status.Error(codes.Internal, "receipt status failed")
+	}
+	return &iotexapi.EstimateActionGasConsumptionResponse{
+		Gas: receipt.GasConsumed,
+	}, nil
+}
+
+func (api *Server) estimateActionGasConsumptionForTransfer(transfer *iotextypes.Transfer) (*iotexapi.EstimateActionGasConsumptionResponse, error) {
+	payloadSize := uint64(len(transfer.Payload))
+	return &iotexapi.EstimateActionGasConsumptionResponse{
+		Gas: payloadSize*action.TransferPayloadGas + action.TransferBaseIntrinsicGas,
```

Comment:
I think you can call transfer.IntrinsicGas() to do this

Reply:
iotextypes.Transfer no such method,have to convert to sealedenvelop